extensions/amp-analytics/0.1/transport.js

```diff
@@ -0,0 +1,105 @@
+/**
+ * Copyright 2015 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {log} from '../../../src/log';
+import {loadPromise} from '../../../src/event-helper';
+
+/** @const {string} */
+const TAG_ = 'AmpAnalytics.Transport';
+
+/**
+ * @param {!Window} win
+ * @param {string} request
+ * @param {!Array<string>} transports
+ */
+export function sendRequest(win, request, transports) {
+  for (let i = 0; i < transports.length; ++i) {
+    if (transports[i] == 'image') {
+      return sendRequestUsingImage(win, request);
+    } else if (transports[i] == 'beacon') {
+      if (sendRequestUsingBeacon(win, request)) {
+        return;
+      }
+    } else if (transports[i] == 'xhrpost') {
+      if (sendRequestUsingXhr(win, request)) {
+        return;
+      }
+    } else {
+      log.warn(TAG_, 'Invalid transport', transports[i]);
+    }
+  }
+  log.warn(TAG_, 'Failed to send request', request);
+}
+
+/**
+ * @param {!Window} win
+ * @param {string} request
+ */
+function sendRequestUsingImage(win, request) {
+  const image = new Image();
+  image.src = request;
+  image.width = 1;
+  image.height = 1;
+  loadPromise(image).then(() => {
+    log.fine(TAG_, 'Sent image request', request);
+  }).catch(() => {
+    log.warn(TAG_, 'Failed to send image request', request);
+  });
+}
+
+/**
+ * @param {!Window} win
+ * @param {string} request
+ * @return {boolean} True if this browser supports navigator.sendBeacon.
+ */
+function sendRequestUsingBeacon(win, request) {
+  if (!win.navigator.sendBeacon) {
+    return false;
+  }
+  win.navigator.sendBeacon(request, '');
+  log.fine(TAG_, 'Sent beacon request', request);
+  return true;
+}
+
+/**
+ * @param {!Window} win
+ * @param {string} request
+ * @return {boolean} True if this browser supports cross-domain XHR.
+ */
+function sendRequestUsingXhr(win, request) {
+  if (!win.XMLHttpRequest) {
+    return false;
+  }
+  const xhr = new win.XMLHttpRequest();
```

Comment:
Why wouldn't XDomainObject get us cookies? Isn't its main purpose?

`xhr.js` main purpose is to polyfill Fetch API which, if supported by the browser, would cover all the cases that we need. We just need to make sure that our polyfill does the right thing, and I think, it's either already there or very close. Currently the `fetch_` method is private since we didn't have a need before to make it public, but we can export it here as needed.

Let's try to do so and see what's our delta. One thing though - if you are using Chrome for testing, it's kind of hard to test since Chrome implements Fetch API natively and so the things would likely appear great immediately but can later fail on some other browser. So, we just need to make sure we test polyfill'd case sufficiently.